controllers/tls_workflow.go

```diff
@@ -133,24 +126,124 @@ func LinkGatewayToClusterIssuerFunc(objs controller.Store) machinery.LinkFunc {
+		for _, l := range linkedPolicies {
+			if string(l.Spec.TargetRef.Name) == g.GetName() && g.GetNamespace() == l.GetNamespace() {
+				return true
+			}
+		}
+
+		return false
+	})
+
+	return lo.Map(linkedGateways, func(item *gwapiv1.Gateway, index int) machinery.Object {
+		return &machinery.Gateway{Gateway: item}
+	})
+}
+
+// Common functions used across multiple reconcilers
+
+func IsTLSPolicyValid(ctx context.Context, s *sync.Map, policy *kuadrantv1alpha1.TLSPolicy) (bool, error) {
+	logger := controller.LoggerFromContext(ctx).WithName("IsPolicyValid")
+
+	store, ok := s.Load(TLSPolicyAcceptedKey)
+	if !ok {
+		logger.V(1).Info("TLSPolicyAcceptedKey not found, policies will be checked for validity by current status")
+		return meta.IsStatusConditionTrue(policy.Status.Conditions, string(gatewayapiv1alpha2.PolicyReasonAccepted)), nil
+	}
+
+	isPolicyValidErrorMap := store.(map[string]error)
+
+	return isPolicyValidErrorMap[policy.GetLocator()] == nil, isPolicyValidErrorMap[policy.GetLocator()]
+}
+
+func GetTLSPoliciesByEvents(topology *machinery.Topology, events []controller.ResourceEvent) []machinery.Policy {
+	policies := lo.Filter(topology.Policies().Items(), func(item machinery.Policy, index int) bool {
+		_, ok := item.(*kuadrantv1alpha1.TLSPolicy)
+		return ok
+	})
+
+	var affectedPolicies []machinery.Policy
+	for _, event := range events {
+		if event.Kind == machinery.GatewayGroupKind {
+			ob := event.NewObject
+			if ob == nil {
+				ob = event.OldObject
 			}
 
-			// Can infer linked gateways through the policy
-			linkedGateways := lo.Filter(gateways, func(g *gwapiv1.Gateway, index int) bool {
-				for _, l := range linkedPolicies {
-					if string(l.Spec.TargetRef.Name) == g.GetName() && g.GetNamespace() == l.GetNamespace() {
+			g := machinery.Gateway{Gateway: ob.(*gwapiv1.Gateway)}
+
+			affectedPolicies = append(affectedPolicies, lo.Filter(policies, func(item machinery.Policy, index int) bool {
+				for _, tg := range item.GetTargetRefs() {
+					if g.GetLocator() == tg.GetLocator() {
 						return true
 					}
 				}
-
 				return false
-			})
+			})...)
+		}
 
-			return lo.Map(linkedGateways, func(item *gwapiv1.Gateway, index int) machinery.Object {
-				return &machinery.Gateway{Gateway: item}
-			})
-		},
+		if event.Kind == kuadrantv1alpha1.TLSPolicyGroupKind {
+			ob := event.NewObject
+			if ob == nil {
+				ob = event.OldObject
+			}
+
+			affectedPolicies = append(affectedPolicies, lo.Filter(policies, func(item machinery.Policy, index int) bool {
+				return item.GetName() == ob.GetName() && item.GetNamespace() == ob.GetNamespace()
+			})...)
+		}
+
+		if event.Kind == CertManagerCertificateKind {
+			ob := event.NewObject
+			if ob == nil {
+				ob = event.OldObject
+			}
+
+			affectedPolicies = append(affectedPolicies, lo.Filter(policies, func(item machinery.Policy, index int) bool {
+				p := item.(*kuadrantv1alpha1.TLSPolicy)
+				return utils.IsOwnedBy(ob, p)
+			})...)
```

Comment:
I believe you could leverage the topology links here too:

```suggestion
			c, _ := ob.(*controller.RuntimeObject).Object.(*certmanagerv1.Certificate)
			listeners := topology.Targetables().Parents(machinery.LocatorFromObject(c))
			affectedPolicies = append(affectedPolicies, lo.FlatMap(listeners, func(l machinery.Targetable) ([]machinery.Policy, bool) {
				return l.(*machinery.Listener).Gateway.Policies()
			})...)
```

In this particular case, not sure if it's better or worse. I'm just thinking about avoiding filtering the same list of policies so many times.

Reply:
Hmm, that's true. If you feel strongly on this I can update, but using the owner refernence looks a lot more readable unless filtering the list of policies should be avoided in general? 🤔

Reply:
I would not say we should avoid filtering policies. Rather, I would say we should avoid writing new logic for things otherwise represented in the topology.

I believe what we want here is get to the list of policies. There must always be a path in the topology between every object directly or indirectly related to a policy and the policy object itself. Getting to the policies some other way suggests to me that we may have failed to represent those links between objects in the topology, that we are underusing the links (and therefore the helper functions such as `Parents` and `Children`, thought to impose a given structure to the topology), or yet that we're using the wrong relationship between objects as topological links (e.g. the object is linked to a resource _and_ it's owned by the resource).

If ownership in this case defines the link between objects, then we should use the link. (Using `IsOwnedBy` would be repeating ourselves.) If the ownership is not the basis for the link, but added only as a mean to facilitate cleaning up resources, then let's not use it as the strong link tying the objects together.

Reply:
Yeah, that's fair, I've reverted this part where this is done, but I'll have a look in the other code changes to see can I better leverage the links 👍